src/dragonball/src/address_space_manager.rs

```diff
@@ -0,0 +1,829 @@
+/// Concrete GuestAddressSpace type used by the VMM.
+pub type GuestAddressSpaceImpl = GuestMemoryAtomic<GuestMemoryMmap>;
+
+/// Concrete GuestMemory type used by the VMM.
+pub type GuestMemoryImpl = <Arc<vm_memory::GuestMemoryMmap> as GuestAddressSpace>::M;
+/// Concrete GuestRegion type used by the VMM.
+pub type GuestRegionImpl = GuestRegionMmap;
+
+// Maximum number of working threads for memory pre-allocation.
+const MAX_PRE_ALLOC_THREAD: u64 = 16;
+
+// Control the actual number of pre-allocating threads. Use one more working thread per 4G memory.
+const PRE_ALLOC_GRANULARITY: u64 = 32;
+
+// Maximal NUMA node ids on host
+const MAX_NODE: u32 = 64;
+
+/// Errors associated with virtual machine address space management.
+#[derive(Debug, thiserror::Error)]
+pub enum AddressManagerError {
+    /// Invalid address space operation.
+    #[error("invalid address space operation")]
+    InvalidOperation,
+
+    /// Invalid address range.
+    #[error("invalid address space region (0x{0:x}, 0x{1:x})")]
+    InvalidAddressRange(u64, GuestUsize),
+
+    /// Failure in initializing guest memory.
+    #[error("failure while initializing guest memory")]
+    MemoryNotInitialized,
+
+    /// No resource available.
+    #[error("no available resource")]
+    NoAvailableResource,
+
+    /// Failed to create memfd to map anonymous memory.
+    #[error("cannot create memfd to map anonymous memory")]
+    CreateMemFd(#[source] nix::Error),
+
+    /// Failed to open memory file.
+    #[error("cannot open memory file")]
+    OpenFile(#[source] std::io::Error),
+
+    /// Failed to set size for memory file.
+    #[error("cannot set size for memory file")]
+    SetFileSize(#[source] std::io::Error),
+
+    /// Failed to unlink memory file.
+    #[error("cannot unlike memory file")]
+    UnlinkFile(#[source] nix::Error),
+
+    /// Failed to duplicate fd of memory file.
+    #[error("cannot duplicate memory file descriptor")]
+    DupFd(#[source] nix::Error),
+
+    /// Failed to create GuestMemory
+    #[error("failure in creating guest memory object")]
+    CreateGuestMemory(#[source] vm_memory::Error),
+
+    /// Failed to mmap() guest memory
+    #[error("cannot mmap() guest memory into current process")]
+    MmapGuestMemory(#[source] vm_memory::mmap::MmapRegionError),
+
+    /// Failure in accessing the memory located at some address.
+    #[error("cannot access guest memory located at 0x{0:x}")]
+    AccessGuestMemory(u64, #[source] vm_memory::mmap::Error),
+
+    /// Failed to set KVM memory slot.
+    #[error("failure while configuring KVM memory slot")]
+    KvmSetMemorySlot(#[source] kvm_ioctls::Error),
+
+    /// Failed to set madvise on AddressSpaceRegion
+    #[error("cannot madvice() on guest memory region")]
+    Madvise(#[source] nix::Error),
+
+    /// join threads fail
+    #[error("join threads fail")]
+    JoinFail,
+
+    /// Failed to access address_space.
+    #[error("failed to access address_space")]
+    AddressSpaceNotInitialized,
```

Comment:
- Actually, there is no difference, so I merge those two errors into one
- The comments could not be deleted since it's necessary to comment every variant for the pub enum or else rust will have warning about that :)